src/server/worker/datum_iterator.go

```diff
@@ -0,0 +1,285 @@
+		Commit:  client.NewCommit(input.Repo, input.Commit),
+		Pattern: input.Glob,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for {
+		fileInfo, err := fs.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, &Input{
+			FileInfo:   fileInfo,
+			Name:       input.Name,
+			Lazy:       input.Lazy,
+			Branch:     input.Branch,
+			EmptyFiles: input.EmptyFiles,
+		})
+	}
+	// We sort the inputs so that the order is deterministic. Note that it's
+	// not possible for 2 inputs to have the same path so this is guaranteed to
+	// produce a deterministic order.
+	sort.Slice(result.inputs, func(i, j int) bool {
+		// We sort by descending size first because it can boost performance to
+		// process the biggest datums first.
+		if result.inputs[i].FileInfo.SizeBytes != result.inputs[j].FileInfo.SizeBytes {
+			return result.inputs[i].FileInfo.SizeBytes > result.inputs[j].FileInfo.SizeBytes
+		}
+		return result.inputs[i].FileInfo.File.Path < result.inputs[j].FileInfo.File.Path
+	})
+	// make sure it gets initialized properly
+	result.Reset()
+	return result, nil
+}
+
+func (d *pfsDatumIterator) Reset() {
+	d.location = -1
+}
+
+func (d *pfsDatumIterator) Len() int {
+	return len(d.inputs)
+}
+
+func (d *pfsDatumIterator) Datum() []*Input {
+	return []*Input{d.inputs[d.location]}
+}
+
+func (d *pfsDatumIterator) Next() bool {
+	d.location++
+	return d.location < len(d.inputs)
+}
+
+type unionDatumIterator struct {
+	iterators []DatumIterator
+	unionIdx  int
+}
+
+func newUnionDatumIterator(pachClient *client.APIClient, union []*pps.Input) (DatumIterator, error) {
+	result := &unionDatumIterator{}
+	for _, input := range union {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.iterators = append(result.iterators, datumIterator)
+	}
+	result.Reset()
+	return result, nil
+}
+
+func (d *unionDatumIterator) Reset() {
+	for _, input := range d.iterators {
+		input.Reset()
+	}
+	d.unionIdx = 0
+}
+
+func (d *unionDatumIterator) Len() int {
+	result := 0
+	for _, datumIterator := range d.iterators {
+		result += datumIterator.Len()
+	}
+	return result
+}
+
+func (d *unionDatumIterator) Next() bool {
+	if d.unionIdx >= len(d.iterators) {
+		return false
+	}
+	if !d.iterators[d.unionIdx].Next() {
+		d.unionIdx++
+		return d.Next()
+	}
+	return true
+}
+
+func (d *unionDatumIterator) Datum() []*Input {
+	return d.iterators[d.unionIdx].Datum()
+}
+
+type crossDatumIterator struct {
+	inputs  []DatumIterator
+	started bool
+}
+
+func newCrossDatumIterator(pachClient *client.APIClient, cross []*pps.Input) (DatumIterator, error) {
+	result := &crossDatumIterator{}
+	for _, input := range cross {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		// start the iterator, and make sure it isn't empty
+		if !datumIterator.Next() {
+			return &crossDatumIterator{}, nil
```

Comment:
I think there is still an issue here without a length check in Next. The first Next call will return true with no inputs because started will be false. I think it might make sense to have the Reset function actually check for a Next in each input and empty the inputs if there is not. So, you could just defer Reset similar to what I mentioned above and remove this check. Then, either add a len(d.inputs) check to Next or set started to true for this edge case.